### PR TITLE
Return `false` for hash_hmac/hash_hmac_file on invalid algorithm

### PR DESCRIPTION
**Pull request by pascalheidmann**
This PR adds a dynamic function return type extension for both `hash_hmac` and `hash_hmac_file` as both functions can return `false` in certain circumstances. For both functions this means providing an invalid algorithm as first parameter will statically return false, otherwise `string` for `hash_hmac` and `string|false` for `hash_hmac_file` (we can't check for file readability in static analysis, and you should always check for file errors anyway). 

There has been a [PR more than 1.5 years ago](https://github.com/phpstan/phpstan/pull/1852) that tried to fix this within static functions map but that would lead to far more false positives.

### CODE REVIEW

File: src/Type/Php/HashHmacFunctionsReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+	/** @var string[] */
+	private array $functions = [
+		'hash_hmac',
+		'hash_hmac_file',
+	];
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return in_array($functionReflection->getName(), $this->functions, true);
```
Comment (ondrejmirtes) [inline src/Type/Php/HashHmacFunctionsReturnTypeExtension.php:28]: This array can be here statically, it's not referenced anywhere else in the extension.

File: src/Type/Php/HashHmacFunctionsReturnTypeExtension.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$algo = $functionCall->args[0]->value;
+		if (!in_array($algo, hash_hmac_algos(), true)) {
```
Comment (ondrejmirtes) [inline src/Type/Php/HashHmacFunctionsReturnTypeExtension.php:34]: There needs to be a fallback for PHP 7.1 using https://www.php.net/manual/en/function.hash-algos.php

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Otherwise this looks great 👍

File: src/Type/Php/HashHmacFunctionsReturnTypeExtension.php (hunk 1 above)
Comment (pascalheidmann) [reply src/Type/Php/HashHmacFunctionsReturnTypeExtension.php]: Have inlined the value, hope this is correct?

I had orientated on other instances like `ArrayPointerFunctionsDynamicReturnTypeExtension` which also uses the `private array $functions = []` notation.

File: src/Type/Php/HashHmacFunctionsReturnTypeExtension.php (hunk 2 above)
Comment (pascalheidmann) [reply src/Type/Php/HashHmacFunctionsReturnTypeExtension.php]: I have added `function_exists('hash_hmac_algos') ? hash_hmac_algos() : hash_algos()` check but sadly PHPStan doesn't care. Any suggestion?

Comment (ondrejmirtes) [reply src/Type/Php/HashHmacFunctionsReturnTypeExtension.php]: It's fine, just make the tests pass and I'll fix the PHPStan issue just before merging :) Thanks.

**Comment by dktapps** [comment]
Just a 2 cents: Maybe it might make more sense to specify the `algo` parameter as a union of all available hash functions as `UnionType(ConstantStringType)` so that this isn't relying on the existence of `false` return to make errors show up in user code.
Context: in PHP 8, this triggers a `ValueError` instead of returning `false`: https://github.com/php/php-src/blob/efe6006ced4f91d0df1fdb0aef7f0079273ac0d0/ext/hash/hash.c#L496

**Comment by pascalheidmann** [comment]
@ondrejmirtes I have replaced the conditional declaration of those supported algorithms with a static one like @dktapps suggested.

As the behaviour seems to change from return `false` to a more convenient Exception driven approach: How should this be handled?

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by ondrejmirtes** [comment]
I'll handle that case when I'll be adding support for PHP 8.

**Pull request merged**